File: lista_dupla_ligada.py
import csv_reader

class Registro:
    """Classe para armazenar os dados de terremotos."""
    def __init__(self, date="", time="", city="", country="", latitude="", 
                 longitude="", earthquake_magnitude="", depth="", impact_score=""):
        self.date = date
        self.time = time
        self.city = city
        self.country = country
        self.latitude = latitude
        self.longitude = longitude
        self.earthquake_magnitude = earthquake_magnitude
        self.depth = depth
        self.impact_score = impact_score

class Node:
    """Nó da lista duplamente encadeada."""
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class ListaDuplaEncadeada:
    """Implementação de uma lista duplamente encadeada."""
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    
    def append(self, data):
        """Adiciona um elemento ao final da lista."""
        new_node = Node(data)
        
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        
        self.size += 1
    
    def remove(self, index):
        """Remove um elemento pelo índice."""
        if index < 0 or index >= self.size:
            return False
        
        current = self.head
        for i in range(index):
            current = current.next
        
        if current.prev:
            current.prev.next = current.next
        else:
            self.head = current.next
        
        if current.next:
            current.next.prev = current.prev
        else:
            self.tail = current.prev
        
        self.size -= 1
        return True
    
    def get_size(self):
        """Retorna o tamanho da lista."""
        return self.size
    
    def is_empty(self):
        """Verifica se a lista está vazia."""
        return self.size == 0
    
    def buscar_por_cidade(self, cidade):
        """Busca registros por nome da cidade."""
        result = []
        current = self.head
        
        while current:
            if cidade.lower() in current.data.city.lower():  # Busca parcial case-insensitive
                result.append(current.data)
            current = current.next
        
        return result
    
    def buscar_por_magnitude(self, magnitude):
        """Busca registros por magnitude exata."""
        result = []
        current = self.head
        
        while current:
            try:
                if float(current.data.earthquake_magnitude) == float(magnitude):
                    result.append(current.data)
            except ValueError:
                pass  # Ignora registros com magnitude não numérica
            current = current.next
        
        return result
    
    def listar_por_magnitude_minima(self, min_magnitude):
        """Lista registros com magnitude maior ou igual à especificada."""
        result = []
        current = self.head
        
        while current:
            try:
                if float(current.data.earthquake_magnitude) >= float(min_magnitude):
                    result.append(current.data)
            except ValueError:
                pass  # Ignora registros com magnitude não numérica
            current = current.next
        
        return result
    
    def get_all(self):
        """Retorna todos os registros da lista."""
        result = []
        current = self.head
        
        while current:
            result.append(current.data)
            current = current.next
        
        return result

def carregar_dataset():
    """Carrega o dataset e converte para a lista duplamente encadeada."""
    import csv_reader
    df = csv_reader.load_data()
    if df is not None:
        return converter_para_lista(df)
    return None

def converter_para_lista(dataframe):
    """Converte um DataFrame pandas para uma ListaDuplaEncadeada."""
    lista = ListaDuplaEncadeada()
    
    for _, row in dataframe.iterrows():
        # Pula linhas totalmente vazias
        if row.isnull().all():
            continue
        
        # Verifica as colunas disponíveis no DataFrame
        date = str(row.get('Date', '')) if 'Date' in row else ''
        time = str(row.get('Time (UTC)', '')) if 'Time (UTC)' in row else ''
        city = str(row.get('City', '')) if 'City' in row else ''
        country = str(row.get('Country', '')) if 'Country' in row else ''
        latitude = str(row.get('Latitude', '')) if 'Latitude' in row else ''
        longitude = str(row.get('Longitude', '')) if 'Longitude' in row else ''
        earthquake_magnitude = str(row.get('Earthquake Magnitude', '')) if 'Earthquake Magnitude' in row else ''
        depth = str(row.get('Depth (km)', '')) if 'Depth (km)' in row else ''
        impact_score = str(row.get('Impact Score', '')) if 'Impact Score' in row else ''
        
        registro = Registro(
            date=date,
            time=time,
            city=city,
            country=country,
            latitude=latitude,
            longitude=longitude,
            earthquake_magnitude=earthquake_magnitude,
            depth=depth,
            impact_score=impact_score
        )
        
        # Só adiciona se tiver pelo menos uma informação relevante
        if any([date, city, country, earthquake_magnitude]):
            lista.append(registro)
    
    return lista

# Para teste direto deste módulo
if __name__ == "__main__":
    print("Carregando dataset para lista duplamente encadeada...")
    lista = carregar_dataset()
    if lista:
        print(f"Dataset carregado com sucesso! Total de registros: {lista.get_size()}")
        
        # Exibir alguns registros para teste
        registros = lista.get_all()
        for i, reg in enumerate(registros[:5]):  # Exibe os primeiros 5 registros
            print(f"\nRegistro #{i}:")
            print(f"Data: {reg.date}")
            print(f"Cidade: {reg.city}")
            print(f"País: {reg.country}")
            print(f"Magnitude: {reg.earthquake_magnitude}")
    else:
        print("Falha ao carregar o dataset.")

File: lista_dupla_ligada.cpp
#include "lista_dupla_ligada.h" // Certifique-se que csv_reader.h está incluído em lista_dupla_ligada.h ou que EarthquakeRecord é conhecido
#include <stdexcept> // Para std::stod, std::invalid_argument, std::out_of_range
#include <iomanip> // Para std::setw, std::left (na formatação da listagem)
#include <iostream> // Para std::cout, std::cerr

// --- Implementação da Classe DoublyLinkedList ---

DoublyLinkedList::DoublyLinkedList() : head(nullptr), tail(nullptr), count(0) {}

DoublyLinkedList::~DoublyLinkedList() {
clear_list();
}

void DoublyLinkedList::clear_list() {
Node* current = head;
while (current != nullptr) {
Node* next_node = current->next;
delete current;
current = next_node;
}
head = nullptr;
tail = nullptr;
count = 0;
}

bool DoublyLinkedList::is_empty() const {
return head == nullptr;
}

int DoublyLinkedList::get_count() const {
return count;
}

// Função privada para exibir um registro (usada internamente por funções de listagem)
void DoublyLinkedList::display_record_details(const EarthquakeRecord& record) const {
// Usando a função global displayRecord de csv_reader.h para consistência,
// ou você pode manter uma formatação específica aqui.
// Se displayRecord de csv_reader.h já faz o que você quer:
std::cout << " "; // Pequena indentação
displayRecord(record); // Chamando a função global (precisa estar declarada em csv_reader.h e incluída)

// Ou, se quiser uma formatação específica para a lista:
/*
std::cout << "  Date: " << std::setw(12) << std::left << record.date
          << "Time (UTC): " << std::setw(10) << std::left << record.time // Assumindo que o campo é 'time'
          << "Magnitude: " << std::setw(5) << std::left << record.magnitude
          << "Depth: " << std::setw(7) << std::left << record.depth + " km" // Assumindo que o campo é 'depth'
          << "Impact: " << std::setw(5) << std::left << record.impact_score << std::endl;
std::cout << "  Location: " << record.city << ", " << record.country
          << " (Lat: " << record.latitude << ", Lon: " << record.longitude << ")" << std::endl;
std::cout << "  --------------------------------------------------------------------------------" << std::endl;
*/
}

// 1. Adicionar novo registro (insere no final)
void DoublyLinkedList::insert_record(const EarthquakeRecord& record) {
Node* new_node = new Node(record);
if (is_empty()) {
head = new_node;
tail = new_node;
} else {
tail->next = new_node;
new_node->prev = tail;
tail = new_node;
}
count++;
}

// 2. Remover registro
// Remove o primeiro registro que corresponder aos critérios de identificação.
bool DoublyLinkedList::remove_record(const std::string& date, const std::string& time_utc, const std::string& city, const std::string& country) {
    if (is_empty()) {
        return false;
    }

    Node* current = head;
    while (current != nullptr) {
        if (current->data.date == date &&
            current->data.time == time_utc &&
            current->data.city == city &&
            current->data.country == country) {

            // Remoção do nó
            if (current == head) {
                head = current->next;
                if (head != nullptr) {
                    head->prev = nullptr;
                } else {
                    tail = nullptr;
                }
            } else if (current == tail) {
                tail = current->prev;
                tail->next = nullptr;
            } else {
                current->prev->next = current->next;
                current->next->prev = current->prev;
            }
            delete current;
            count--;
            return true;
        }
        current = current->next;
    }
    return false;
}

// 3. Funções de Busca
std::vector<EarthquakeRecord> DoublyLinkedList::search_by_city(const std::string& city_name) const {
std::vector<EarthquakeRecord> results;
Node* current = head;
while (current != nullptr) {
if (current->data.city == city_name) {
results.push_back(current->data);
}
current = current->next;
}
return results;
}

std::vector<EarthquakeRecord> DoublyLinkedList::search_by_magnitude_exact(const std::string& magnitude_value) const {
std::vector<EarthquakeRecord> results;
Node* current = head;
while (current != nullptr) {
if (current->data.magnitude == magnitude_value) {
results.push_back(current->data);
}
current = current->next;
}
return results;
}

std::vector<EarthquakeRecord> DoublyLinkedList::search_by_date(const std::string& date_value) const {
std::vector<EarthquakeRecord> results;
Node* current = head;
while (current != nullptr) {
if (current->data.date == date_value) {
results.push_back(current->data);
}
current = current->next;
}
return results;
}

std::vector<EarthquakeRecord> DoublyLinkedList::search_by_city_and_magnitude(const std::string& city_name, const std::string& magnitude_value) const {
std::vector<EarthquakeRecord> results;
Node* current = head;
while (current != nullptr) {
if (current->data.city == city_name && current->data.magnitude == magnitude_value) {
results.push_back(current->data);
}
current = current->next;
}
return results;
}

// 4. Funções de Listagem
void DoublyLinkedList::list_all_records() const {
if (is_empty()) {
std::cout << "A lista de registros de terremotos esta vazia." << std::endl;
return;
}
std::cout << "\n--- Todos os Registros de Terremotos na Lista (" << count << ") ---" << std::endl;
Node* current = head;
int record_num = 1;
while (current != nullptr) {
std::cout << "Registro #" << record_num++ << ":" << std::endl;
display_record_details(current->data); // Usa a função privada ou a global
if (record_num > 1) std::cout << " --------------------------------------------------------------------------------" << std::endl; // Para separar melhor
current = current->next;
}
}

void DoublyLinkedList::list_by_min_magnitude(const std::string& min_mag_str) const {
if (is_empty()) {
std::cout << "A lista esta vazia. Nada para listar." << std::endl;
return;
}

double min_magnitude_val;
try {
    min_magnitude_val = std::stod(min_mag_str);
} catch (const std::invalid_argument& ia) {
    std::cerr << "Erro: Valor de magnitude minima invalido: " << min_mag_str << std::endl;
    return;
} catch (const std::out_of_range& oor) {
    std::cerr << "Erro: Valor de magnitude minima fora do intervalo: " << min_mag_str << std::endl;
    return;
}

std::cout << "\n--- Registros com Magnitude >= " << min_mag_str << " ---" << std::endl;
Node* current = head;
int found_count = 0;
int record_num = 1;
while (current != nullptr) {
    try {
        double current_mag = std::stod(current->data.magnitude);
        if (current_mag >= min_magnitude_val) {
            std::cout << "Registro Filtrado #" << record_num++ << ":" << std::endl;
            display_record_details(current->data);
            if (found_count > 0) std::cout << "  --------------------------------------------------------------------------------" << std::endl;
            found_count++;
        }
    } catch (const std::exception& e) {
        // Ignorar silenciosamente ou logar se um registro tiver magnitude mal formatada
    }
    current = current->next;
}
if (found_count == 0) {
    std::cout << "Nenhum registro encontrado com magnitude >= " << min_mag_str << "." << std::endl;
}
}

void DoublyLinkedList::list_by_year(const std::string& year_str) const {
if (is_empty()) {
std::cout << "A lista esta vazia. Nada para listar." << std::endl;
return;
}
if (year_str.length() != 4) {
std::cerr << "Erro: Formato de ano invalido. Use AAAA (ex: 2023)." << std::endl;
return;
}
try {
std::stoi(year_str);
} catch (const std::exception& e) {
std::cerr << "Erro: Ano invalido: " << year_str << std::endl;
return;
}

std::cout << "\n--- Registros do Ano " << year_str << " ---" << std::endl;
Node* current = head;
int found_count = 0;
int record_num = 1;
while (current != nullptr) {
    if (current->data.date.rfind(year_str, 0) == 0) {
        std::cout << "Registro Filtrado #" << record_num++ << ":" << std::endl;
        display_record_details(current->data);
        if (found_count > 0) std::cout << "  --------------------------------------------------------------------------------" << std::endl;
        found_count++;
    }
    current = current->next;
}
if (found_count == 0) {
    std::cout << "Nenhum registro encontrado para o ano " << year_str << "." << std::endl;
}
}

// Função para retornar todos os registros como um vetor
// Útil para operações externas que precisam de todos os dados (ex: reclassificação)
std::vector<EarthquakeRecord> DoublyLinkedList::get_all_records_vector() const {
std::vector<EarthquakeRecord> all_records;
Node* current = head;
while (current != nullptr) {
all_records.push_back(current->data);
current = current->next;
}
return all_records;
}